design/007-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,421 @@
+
+```shell
+renku dataset add --external ../../path/to/my/data
+```
+
+The path to the data is stored in the dataset's metadata; the pointer can also
+be updated in case that the data changes and the hash needs to be recomputed,
+for the purposes of reproducibility. Of course, the obvious drawback here is
+that the data location is arbitrary - if the project is brought to another
+environment/machine/cluster, it's fairly likely that the data path will be
+incorrect and therefore all recorded workflows will fail.
+
+#### S3
+
+Using S3 in Renku is currently limited to interactive sessions. When launching a
+session, users can specify any S3-compatible URL together with credentials, if
+necessary. The S3 bucket is then mounted in their interactive session.
+Currently, no metadata about these mounts is recorded. The data could, in
+principle, be added to a renku dataset with the "external" mechanism described
+above, but no metadata about its actual origins (i.e. an S3 bucket) would be
+recorded.
+
+#### Upcoming
+
+We have plans to include other external data providers, such as NFS shares,
+openBIS etc. From the user's perspective, they all will generally work in the
+same way as S3; data will be mounted to a local directory outside the repository
+and made accessible as network-mounted storage.
+
+## Design Detail
+
+Generically, we should strive to remove the separation between "hosted" Renku
+(i.e. what I can do if I run an interactive session on renkulab.io) and
+"offline" Renku (what I can do with the CLI anywhere).
+
+Ideally we could once again follow the plug-in paradigm in the CLI, like we
+do for other functionality.
+
+### Provider interface
+
+If we adopt the view that data access should be done on the filesystem level,
+then one approach would be to rely on FUSE as the interface. This means using
+existing FUSE implementations where possible (e.g. S3) or implementing new ones
+using e.g. [`python-fuse` library](https://github.com/libfuse/python-fuse).
+It is not obvious that relying on FUSE is a good idea at all -- we need to conduct
+a study of pros/cons before committing to this path. There might be cases where
+FUSE would not be universally acceptable - there, we should offer the option to
+use an alternative, e.g. copy instead of stream the data.
+
+The interface should rely on a `DataSource` entity type. The role of the
+Provider is to configure the project and/or the user's environment in such a way
+as to make the `DataSource` available in a consistent, repeatable location with
+respect to the project and without boilerplate.
+
+### Local vs. hosted interactive sessions
+
+Relying on FUSE for providing filesystem access to external resources means that
+locally we do not require special privileges and can manage the data access for
+the user. In the docker-based remote sessions, the situation is more complicated
+because containers require elevated privileges to access FUSE on the host.
```

Comment:
This could be a problem for hardened/enterprise deployments that have strict security policies that are not allowed to be changed or require a lot of effort so that we can get permission to deviate from them.

Reply:
yes - dynamic mounting might have to be a feature that you can turn off if needed.